internal/core/src/segcore/Utils.cpp

```diff
@@ -0,0 +1,258 @@
+            obj->mutable_data()->Add(data, data + count);
+            break;
+        }
+        case DataType::DOUBLE: {
+            auto data = reinterpret_cast<const double*>(data_raw);
+            auto obj = scalar_array->mutable_double_data();
+            obj->mutable_data()->Add(data, data + count);
+            break;
+        }
+        case DataType::VARCHAR: {
+            auto data = reinterpret_cast<const std::string*>(data_raw);
+            auto obj = scalar_array->mutable_string_data();
+            for (auto i = 0; i < count; i++) *(obj->mutable_data()->Add()) = data[i];
+            break;
+        }
+        default: {
+            PanicInfo("unsupported datatype");
+        }
+    }
+
+    return data_array;
+}
+
+std::unique_ptr<DataArray>
+CreateVectorDataArrayFrom(const void* data_raw, int64_t count, const FieldMeta& field_meta) {
+    auto data_type = field_meta.get_data_type();
+    auto data_array = std::make_unique<DataArray>();
+    data_array->set_field_id(field_meta.get_id().get());
+    data_array->set_type(milvus::proto::schema::DataType(field_meta.get_data_type()));
+
+    auto vector_array = data_array->mutable_vectors();
+    auto dim = field_meta.get_dim();
+    vector_array->set_dim(dim);
+    switch (data_type) {
+        case DataType::VECTOR_FLOAT: {
+            auto length = count * dim;
+            auto data = reinterpret_cast<const float*>(data_raw);
+            auto obj = vector_array->mutable_float_vector();
+            obj->mutable_data()->Add(data, data + length);
+            break;
+        }
+        case DataType::VECTOR_BINARY: {
+            AssertInfo(dim % 8 == 0, "Binary vector field dimension is not a multiple of 8");
+            auto num_bytes = count * dim / 8;
+            auto data = reinterpret_cast<const char*>(data_raw);
+            auto obj = vector_array->mutable_binary_vector();
+            obj->assign(data, num_bytes);
+            break;
+        }
+        default: {
+            PanicInfo("unsupported datatype");
+        }
+    }
+    return data_array;
+}
+
+std::unique_ptr<DataArray>
+CreateDataArrayFrom(const void* data_raw, int64_t count, const FieldMeta& field_meta) {
+    auto data_type = field_meta.get_data_type();
+
+    if (!datatype_is_vector(data_type)) {
+        return CreateScalarDataArrayFrom(data_raw, count, field_meta);
+    }
+
+    return CreateVectorDataArrayFrom(data_raw, count, field_meta);
+}
+
+// TODO remove merge dataArray, instead fill target entity when get data slice
+std::unique_ptr<DataArray>
+MergeDataArray(std::vector<std::pair<milvus::SearchResult*, int64_t>>& result_offsets, const FieldMeta& field_meta) {
+    auto data_type = field_meta.get_data_type();
+    auto data_array = std::make_unique<DataArray>();
+    data_array->set_field_id(field_meta.get_id().get());
+    data_array->set_type(milvus::proto::schema::DataType(field_meta.get_data_type()));
+
+    for (auto& result_pair : result_offsets) {
```

Comment:
should not add 1 element each time.
Reserve result_offsets.size to avoid copy in add

Reply:
I'll change this in next pr.